rmw_connextdds_common/src/common/rmw_impl.cpp

```diff
@@ -1371,17 +1379,24 @@ RMW_Connext_Subscriber::finalize(const bool reset_cft)
     }
   }
 
-  if (nullptr != this->dds_reader && DDS_RETCODE_OK !=
+  if (nullptr == this->dds_reader) {
```

Comment:
I would move this check before the previous block, because logically "having a DDS data reader" is also a precondition for calling `this->return_messages()`, not just for `this->dds_participant()` (good catch on that).

Also, maybe change the error message to something a bit more specific, like "no DDS DataReader to finalize".